### PR TITLE
Fixing review state to APPROVED whe 'LGTM' in COMMENTED review

### PR DESCRIPTION
**Pull request by Tiriel**
From #67 
Tested with nodejs/node#16248 as given in issue #67 

Ran lint and tests.

Comments and reviews deeply welcomed!

### CODE REVIEW

**Comment by cPhost** [comment]
@Tiriel Great Work! can you add tests in `test/unit/reviews.test.js`

**Comment by Tiriel** [comment]
Oh yeah sorry, that's indeed a good idea. I'm on it!

**Comment by Tiriel** [comment]
Original tests were in fact all that was needed. I just had to add a test-case in fixtures for my first draft to make some tests fail. (pr_data tests in fact).

Corrected work, now tests still pass with proper test case, and command still responds properly.

File: lib/reviews.js [hunk 1]
```diff
@@ -132,14 +138,31 @@ class ReviewAnalyzer {
...
+    return review.state === COMMENTED && this.hasLGTM(review, 'ref');
+  }
+
+  /**
+   * @param object
+   * @param prop: string
+   * @returns {boolean}
+   */
+  hasLGTM(object, prop) {
+    return LGTM_RE.test(object[prop]);
```
Comment (refack) [inline lib/reviews.js]: So this is too loose. It would pass for `I'm revoking my LGTM`.
IMHO this should be `object[prop].toLowerCase() === 'lgtm'`.

Comment (joyeecheung) [reply lib/reviews.js]: @refack This is already the case for normal comments, I think we should log an info here, like what we do for those, to remind the landers to double-check. The actual context would be too hard to parse without some kind of natural language processing..

Comment (refack) [reply lib/reviews.js]: Good idea!

Comment (joyeecheung) [reply lib/reviews.js]: OK not log an info here, but make a new type of `REVIEW_SOURCES`, maybe `FROM_REVIEW_COMMENT`, and warn them in `checkReviews` like what we do for `FROM_COMMENT`

Comment (joyeecheung) [reply lib/reviews.js]: @refack re: `I'm revoking my LGTM.` they would've put an red cross there I think?

Comment (refack) [reply lib/reviews.js]: I'm thinking of something I saw last week - https://github.com/nodejs/node/pull/16612#pullrequestreview-73478662

Comment (refack) [reply lib/reviews.js]: That case was not a "Request Changes" , just "Don't count me as approver", same as dismissed review.

**Comment by refack** [comment]
BTW: people can also dismiss other's reviews, so need to keep track of the author attribution

**Comment by joyeecheung** [comment]
@refack Yes, we probably need to dig a bit more into the state machine later as well..

File: lib/reviews.js (hunk 1 above)
Comment (joyeecheung) [reply lib/reviews.js]: Another idea: if we are uncertain about a LGTM, we should log it as `?Reviewed-By: xxx`, and then core-validate-commit should complain about it.

**Comment by Tiriel** [comment]
By the way, I see that the info telling one PR was approved in comments or commented review is on `console.info`. Wouldn't it be better to pass it to `console.warn`?

**Comment by Tiriel** [comment]
By the way, I see that the info telling one PR was approved in comments or commented review is on `console.info`. Wouldn't it be better to pass it to `console.warn`?

**Comment by Tiriel** [comment]
For now, that's what I got:
image

**Comment by Tiriel** [comment]
Okay, I've read all the comments, bit confusing xD

I think I'll go for both things you suggested if possible @joyeecheung , new source and `?`in report.

@joyeecheung @refack Otherwise, what about changing the RegEx to this: `^(L|S)GTM!?$` ? This would prevent most cases of matches when people are discussing "LGTM" without actually wanting to validate a PR and still allowing some leeway, don't you think?

**Comment by apapirovski** [comment]
I know I'm going to be a minority on this but I would scrape the whole LGTM checking. Everyone uses the green button. If they're not using and they say 'LGTM' or something that extent, it usually means they scanned it but don't have complete confidence to give it firm approval.

**Comment by apapirovski** [comment]
For example, what does this count as?

> Basically LGTM, just note that automatic fixing can introduce bugs if another variable of the same name was declared within the function.

I don't think the author would expect to be listed in `Reviewed-By` even though they gave a LGTM.

**Comment by Tiriel** [comment]
That's a good point, but actually that's why I would put the notice in warning.

In a way, just to say "Some people seem to approve, but be careful".

**Comment by joyeecheung** [comment]
Or....maybe we should skip those people in the actual metadata, just log them as "potential reviewers"?

**Comment by apapirovski** [comment]
@joyeecheung I think that's my favourite suggestion so far. I think any LGTM couldbe be included as ```
Approved: 1, 1 from TSC (mhdawson)
N
```

**Comment by apapirovski** [comment]
Fat fingered the close button. Sorry.

**Comment by Tiriel** [comment]
Maybe then keeping the line "XXX reviewed in comment" while removing the review from review count?

**Comment by joyeecheung** [comment]
@Tiriel Yes that's what I had in mine.

**Comment by joyeecheung** [comment]
@Tiriel Or maybe..

```
The following collaborators are not listed in the metadata:
Foo <foo@example.org>: I am not sure about if I can give a proper LGTM but this is a great idea
Bar <bar@example.org>: LGTM modulo style nits
```

(we can even make this interactive and let the user decide!...although for the moment I think this PR is good with just removing them in the metadata)

**Comment by Tiriel** [comment]
Form now that's what I got:
image
Is that okay?

**Comment by Tiriel** [comment]
I can do like you said in your last comment, but I feel that would need a more important rework of the workflow that would justify another PR

**Comment by Tiriel** [comment]
Actually this way other tests fail.

Going to rework already :)

**Comment by Tiriel** [comment]
Hey!

Finally reworked a bit.
I needed to adapt the tests, since the LGTM in comments don't count as approvals now.
The result looks like this:
image

Hope it's ok now!

File: lib/pr_checker.js [hunk 2]
```diff
@@ -77,7 +77,7 @@ class PRChecker {
 
   checkReviews() {
     const {
-      pr, logger, reviewers: { rejected, approved }
+      pr, logger, reviewers: { rejected, approved, commentApproved }
```
Comment (joyeecheung) [inline lib/pr_checker.js]: I don't think we need the extra export here, the information is already in `approved[i].review.source`, so we can derive data from the source of truth here.

File: lib/pr_checker.js [hunk 3]
```diff
@@ -95,8 +95,12 @@ class PRChecker {
       status = false;
       logger.warn(`Approvals: 0`);
```
Comment (joyeecheung) [inline lib/pr_checker.js]: We can build something like `approvalFromReview` and use that from here, so we don't need to do the filtering in `getTSCHint`.

File: lib/pr_checker.js [hunk 4]
```diff
@@ -95,8 +95,12 @@ class PRChecker {
...
     } else {
+      let notComm = approved.length;
+      approved.map((r) => {
+        if (r.review.source !== FROM_REVIEW) { notComm--; }
+      });
       let hint = this.getTSCHint(approved);
-      logger.info(`Approvals: ${approved.length}${hint}`);
+      logger.info(`Approvals: ${notComm}${hint}`);
 
       for (const { reviewer, review } of approved) {
```
Comment (joyeecheung) [inline lib/pr_checker.js]: This can be merged with `review.source === FROM_REVIEW_COMMENT`

File: lib/reviews.js (hunk 1 above)
Comment (joyeecheung) [reply lib/reviews.js]: On a second thought, I am finding the `object[prop].toLowerCase() === 'lgtm'` approach better (although we can just do that by changing `LGTM_RE` to `/^lgtm$/i` and trim the comments before testing them), it's just simpler that way. I think people who don't give LGTM either this way or with a green review would not mind not being listed in the review-by list.

File: lib/pr_checker.js (hunk 2 above)
Comment (Tiriel) [reply lib/pr_checker.js]: Well, I thought so at first, but that's what was breaking the tests, and would have need it to revise them completely.
Plus, IMHO, since I've separated things, I think it's semantically weird to mix properly approved reviews and commented reviews with LGTM. I mean, that's the whole point of getting these reviews out of the metadatas.

File: lib/pr_checker.js (hunk 4 above)
Comment (Tiriel) [reply lib/pr_checker.js]: You mean the two blocks nearly repeating themselves? It probably needs refactoring yes, but I wanted advice on the workflow yet :)

File: lib/pr_checker.js (hunk 3 above)
Comment (Tiriel) [reply lib/pr_checker.js]: Going to look into this!

**Comment by joyeecheung** [comment]
@Tiriel Sorry for so much back-and-forth, but I actually kinda see the point of https://github.com/joyeecheung/node-core-utils/pull/72#discussion_r148944765 now, we might as well just ignore the reviews that is not just `^lgtm$/i`, in that case we don't really need to warn about anything because it's pretty clear that those are legitimate sign-offs.

**Comment by Tiriel** [comment]
No prob! I'll try and rollback to the previous version then.
What do you prefer? `r.bodyText.toLowerCase() === 'lgtm'`, or changing the regex to `/^lgtm\W?$/gmi`?

Personnally I prefer modifying the regex, I've seen people adding punctuation to LGTMs and making a strict equal comparison would count them.

**Comment by joyeecheung** [comment]
@Tiriel Let's go with regexp then. Also we might want to trim the text before testing them as well.

**Comment by Tiriel** [comment]
Done!
Linted and tested as usual (now) :D

File: lib/reviews.js [hunk 5]
```diff
@@ -80,6 +81,12 @@ class ReviewAnalyzer {
             new Review(r.state, r.publishedAt, r.url, FROM_REVIEW)
           );
           break;
+        case COMMENTED:
+          map.set(
+            login,
+            new Review(r.state, r.publishedAt, r.bodyText, FROM_REVIEW_COMMENT)
```
Comment (joyeecheung) [inline lib/reviews.js]: I suggest we just pass `APPROVED` as the state here, like what we do for approval in normal comments. This way we don't need to call `isApprovedInComment` in `getReviewers`. We still keep the info about its source in `review.source`.

File: test/fixtures/reviews_approved.json [hunk 6]
```diff
@@ -36,6 +36,15 @@
   "publishedAt": "2017-10-24T14:49:52Z"
 },
 {
+  "bodyText": "LGTM",
+  "state": "COMMENTED",
+  "author": {
+    "login": "Baz"
```
Comment (joyeecheung) [inline test/fixtures/reviews_approved.json]: Can you make a new case for this? This and the case above are overlapping each other so one of them would not be visible in the tests. To do this we need to add another collaborator in `collaborators.json` and the readme in the fixtures as well.

**Comment by joyeecheung** [comment]
Also with this we can probably drop the check for `FROM_COMMENT` in the pr checker, or put it behind a command line flag and warn about both `FROM_COMMENT` and `FROM_REVIEW_COMMENT` (but I don't think we really need a warning for that any more..)

File: lib/reviews.js (hunk 5 above)
Comment (Tiriel) [reply lib/reviews.js]: Oh thanks, that's what I had done in the first place, forgot to reinstate it. when reverting

File: test/fixtures/reviews_approved.json (hunk 6 above)
Comment (Tiriel) [reply test/fixtures/reviews_approved.json]: Sure thing!

Comment (Tiriel) [reply test/fixtures/reviews_approved.json]: There are in fact numerous tests using each files. I also had to add an entry in reviewers_approved.json and modify accordingly other tests. Push coming

**Comment by Tiriel** [comment]
There, this should do it. I added the `--comments` flag and some tests cases. I'll let you review, since I had to adapt some existing tests cases

File: lib/args.js [hunk 7]
```diff
@@ -37,6 +37,12 @@ function buildYargs(args = null) {
       describe: 'File to write the metadata in',
       type: 'string'
     })
+    .option('comments', {
```
Comment (joyeecheung) [inline lib/args.js]: I'd rather let this be `--check-comments` with no aliases, in case we need the `c` later. This does not seem to be very in demand.

File: test/unit/pr_checker.test.js [hunk 8]
```diff
@@ -80,12 +80,13 @@ describe('PRChecker', () => {
 
       const expectedLogs = {
         warn: [
+          ['Quux User(Quux)) approved in via LGTM in comments'],
```
Comment (joyeecheung) [inline test/unit/pr_checker.test.js]: There is a redundant `)` here..

Comment (Tiriel) [reply test/unit/pr_checker.test.js]: Actually this redundant parens is in every test and is in fact in  the base template :D
I'll remove it everywhere.

File: lib/args.js (hunk 7 above)
Comment (Tiriel) [reply lib/args.js]: Your desires are my commands!

**Comment by Tiriel** [comment]
Fixed!

**Comment by cPhost** [comment]
@Tiriel can you add this option in readme.

**Comment by Tiriel** [comment]
Very good idea! Incoming!

File: lib/args.js [hunk 9]
```diff
@@ -37,6 +37,11 @@ function buildYargs(args = null) {
       describe: 'File to write the metadata in',
       type: 'string'
     })
+    .option('check-comments', {
+      demandOption: false,
+      describe: 'Check for\'LGTM\' in comments',
```
Comment (cPhost) [inline lib/args.js]: Same as above^

File: README.md [hunk 10]
```diff
@@ -52,11 +52,12 @@ get-metadata <identifier>
...
-  --version    Show version number                                     [boolean]
-  --owner, -o  GitHub owner of the PR repository                        [string]
-  --repo, -r   GitHub repository of the PR                              [string]
-  --file, -f   File to write the metadata in                            [string]
-  --help, -h   Show help                                               [boolean]
+  --version         Show version number                                [boolean]
+  --owner, -o       GitHub owner of the PR repository                   [string]
+  --repo, -r        GitHub repository of the PR                         [string]
+  --file, -f        File to write the metadata in                       [string]
+  --check-comments  Check for'LGTM' in comments                        [boolean]
```
Comment (cPhost) [inline README.md]: Minor Nit:  add a space between for and LGTM `'Check for \'LGTM\' in comments'`

**Comment by Tiriel** [comment]
Going to merge then, unless someone objects!
